### ISSUE TITLE
German version? :)

### ISSUE THREAD

**Issue by Eikju** [Immemor/RaidCore#1]
Hey there :)

I just translated the English version of "Phageborn Convergence" for my raid because your version won't work for German clients.

Is it possible to work with you for a full translation? What I need is just the English skills, spells or boss names which I should translate.

Thank you very much :)

**Comment by NielsH** [comment]
Hey,

Fun to see some other people finding this addon. Did someone in Modified link it to you?

Eventually having support for multiple client languages is something I'd like to do, though atm I just have very little time even to just build current modules for bosses we are progressing on in DS :p 

If you found the Convergence module I'm sure you found that all the bosses are in the Modules/ directory, translation wise it's not hard to add other languages, either by having a lookup table (which may be the better way of doing it) or just checking for `if castName == "englishName" or castName == "germanName" then`

Do you know any lua? If so, feel free to make a pull request so I can merge in the changes.

Some modules use splIds instead of names, which is better since it works cross-clients language wise. Downside is that it requires some more effort to get the spellids (has to be done during the fight by some debug logging of all the spells)

Detecting datachron messages may be more tricky, it will probably require an actual translation, ids cannot be used here.

If you don't know any lua, just supply me with the ability names/chat messages done by the bosses that are relevant (and perhaps the boss names, I'm not sure if they are any different across clients) and I can add them in.

Hope that helps :+1:

**Comment by Eikju** [comment]
Hey there.

Thank you very much for your help. I'm sorry to must tell you, that I do not anything about lua :D. What I can do is to translate each Boss with the skills and send you the .lua files.

Will that help you to implement it as a standard?

Thank you very much :D

**Comment by NielsH** [comment]
Sure, if you can provide me with the exact translation strings (which are all tested since I cannot do that myself) I can merge it in.

Just make sure to include everything required (boss names, ability names, datachron/NPC_SAY messages), etc... 

Cheers :+1:

**Comment by Eikju** [comment]
I will do my best Best :)

Just one question: How do you find out the NPC_SAY? I'm right now on Kuralak the Defiler.

Tank you :)

**Comment by NielsH** [comment]
For Kuralak the following module is used: https://github.com/NielsH/RaidCore/blob/master/Modules/Kuralak.lua 

The NPC_SAY messages are found here: https://github.com/NielsH/RaidCore/blob/master/Modules/Kuralak.lua#L87

So what I will need is the german version of those strings - the message.find() part.

**Comment by Eikju** [comment]
Yeah :D But where do you find out that she is saying "One of us... you will become one of us"? Do you have access to a specific Wildstar lib or is it just a collection of some wipes where you have noted these lines on a paper? :)

Hopefully you know what I mean.

**Comment by NielsH** [comment]
The only way to know is by doing the fights and screenshotting/chatlogging everything that happens. Then after the fight it can be analyzed to figure out what happens and which strings have to be used in the addon.

No fancy/handy lookup for that sadly :(

**Comment by Eikju** [comment]
Gnaaaah, okay.

But thx for the reply. I will do my best to get these strings :D

**Comment by Eikju** [comment]
So where should I upload the German files? I have translated Genetic Archives completely. I tested it in 2 full raids and it work's like charming :)

**Comment by NielsH** [comment]
It's probably best if you add them as a gist and link that to me, unless you know lua and can add a pull request.

Then I'll merge it in when I can :+1:

**Comment by Eikju** [comment]
Here are the files :)

https://gist.github.com/Eikju/a4c1a380b57187e78831
https://gist.github.com/Eikju/ae69bdb2f463ac17438f
https://gist.github.com/Eikju/befa63b49ed1d0e70211
https://gist.github.com/Eikju/27357f895f63636ff754
https://gist.github.com/Eikju/7523cb665e51ac88286f
https://gist.github.com/Eikju/1eefe94be745355082da

**Comment by NielsH** [comment]
Hey, cool. I'll add them when I can, quite busy atm so not sure how soon it will be!

Once I do add it I'll probably send you an experimental version so you can confirm it works properly :)

**Comment by NielsH** [comment]
I have added a translation branch to test this, with the initial changes for X89 here. However it is untested, could you check if this method of managing the translation strings work? (so just X89 changed for now)

https://github.com/NielsH/RaidCore/commit/7757fab03547d4f4cde5581bc5763653fb12cd24
Branch : https://github.com/NielsH/RaidCore/tree/translation

**Comment by Eikju** [comment]
Thx for the fast integration in your system. I will test it tommorow on our raid. If it will work, I will inform you.

**Comment by NielsH** [comment]
Any luck with this?

**Comment by Eikju** [comment]
Ah, damn. I tought, I have allready answered you the question. It won't work. The Addon doesn't load, if I will use your new tree.

**Comment by NielsH** [comment]
Can you define the issue better? Any errors? Does the addon not load at all? or does the module load but it doesn't start?

**Comment by Eikju** [comment]
The whole Addon (not just the Module!) will not load, if I use your new file. There aren't any errors at all.

**Comment by NielsH** [comment]
Hey,

Sadly if you are unable to debug it there's not much I can do. I can test if it works for me next time I do GA, but I can't say when that will be... Perhaps someone in your guild that knows lua can help you debug it live if you want it done faster?

**Comment by Eikju** [comment]
I can test it again next week, if you tell me how I can see some lua-logs. So maybe I can fix the problem :)

**Comment by NielsH** [comment]
You'll either get them in the debug chat, as a pop-up, or via ESC -> Addons
overview (icon will be yellow or red instead of green, and you can click it
to show the errors). Good luck :)

**Comment by Eikju** [comment]
...ppData\Roaming\NCSOFT\WildStar\Addons\RaidCore\Modules/ExperimentX89.lua:4: unexpected symbol near '<'

So that is the bug :)

**Comment by NielsH** [comment]
Could you check your local X89 module? Looking at https://raw.githubusercontent.com/NielsH/RaidCore/translation/Modules/ExperimentX89.lua - I don't see the '<' symbol anywhere in that file.

**Comment by Eikju** [comment]
That's the error message:
...T\WildStar\Addons\RaidCore\Modules/ExperimentX89.lua:13: attempt to call method 'GetLocale' (a nil value)
stack trace:
    ...T\WildStar\Addons\RaidCore\Modules/ExperimentX89.lua:13: in main chunk

**Comment by NielsH** [comment]
GetLocale should be called on Line 18, not 13. Are you sure you have the full/correct branch? Be aware it's not just the X89 file that changed. The RaidCore.lua file changed as well.

**Comment by Eikju** [comment]
I just copy-pasted the whole ExperimentX89.lua over the old file: https://github.com/NielsH/RaidCore/blob/translation/Modules/ExperimentX89.lua

I didn't replaced the RaidCore.lua. So let me test it :)

**Comment by NielsH** [comment]
The best thing is to just replace all of RaidCore with the translation branch. That's what it's for, and it will prevent conflicts ;)

Make sure to double-check the X89 lua file as well. It says there is a call to GetLocale() on line 13, but as you can see https://github.com/NielsH/RaidCore/blob/translation/Modules/ExperimentX89.lua#L13 here there is no such call on that line. So I suspect something may be wrong there as well.

**Comment by Eikju** [comment]
Jep. Now it works as intended :D. So I can test it next tuesday :)

**Comment by NielsH** [comment]
Awesome! :+1:

**Comment by Eikju** [comment]
Good news @NielsH - it worked like intended!

**Comment by Eikju** [comment]
So can you implement the German translation into the other files? It would be really nice :)

**Comment by NielsH** [comment]
I'll add it when I can, but I can't give you a timeframe. Before I merge them into the main branch I'll poke you to test the other modules. If all is good we can merge it in!

**Comment by Eikju** [comment]
Roger that :)

**Comment by NielsH** [comment]
I've added Kuralak to the translation branch. By the way, did the Big Bomb warning work for X89? It should give an audiovisual message when you get the big bomb.

Also, are you sure you don't have anyone in the guild that knows lua and wants to help out? It would get you the german translations merged in much quicker :p

**Comment by Eikju** [comment]
Thx 4 adding Kuralak! Yes. The warning works for X89. It also worked as I edited the whole English text into German one. I hope you mean the audio file "run away little girl, run away" an so on :dancer: 

And yes, we do not have any lua programmers at our guild. Sorry bro :-(. But would you have thought from the beginning about the translation, then it would be much easier to insert them. Right now it is much more work ;-)

**Comment by NielsH** [comment]
Ya all just need to use a real client language instead of silly french/german :p

**Comment by Eikju** [comment]
Aaaaah... I see what you did there :-þ.

BTW: Where is the new Kuralak file? Maybe I can learn the syntax by my self so I can help you.

**Comment by NielsH** [comment]
It's part of the translation branch - See https://github.com/NielsH/RaidCore -> Select the branch "translation" and then download (right side download as zip)

**Comment by Eikju** [comment]
So if I edit these whole branch, I can be sure that it is the newest version of the English addon, right?

**Comment by NielsH** [comment]
Yeah, there are some changes in the master branch for the bosses, but nothing special (just fixes for better boss detection/activation/deactivation). We can just merge them together once everything is done.

Alternatively you could edit the files from the master branch and submit them to the translation one. It doesn't really matter in the end ;)

**Comment by Eikju** [comment]
Ok. Thank you for the information :)

**Comment by Eikju** [comment]
Ok... I worked 3 hours and modified all GA-Files.
I can test it this raid-week. Here are the files, m8 :)

Phage Maw: https://gist.github.com/Eikju/3284ed3c75206d301c3d
Prototypes: https://gist.github.com/Eikju/8d0249767b07cad0594f
Phage Council: https://gist.github.com/Eikju/c638f314f2f1f33107ba
Ohmna: https://gist.github.com/Eikju/4612f6ffcf672759c3d5

Puh.... :D

**Comment by Eikju** [comment]
Damn.. some Error occured:

...AppData\Roaming\NCSOFT\WildStar\Addons\RaidCore\Modules/PhageCouncil.lua:35: '}' expected (to close '{' at line 19) near 'shieldWarn'

xDDD

**Comment by NielsH** [comment]
You forgot a comma ;) https://gist.github.com/Eikju/c638f314f2f1f33107ba#file-gistfile1-txt-L34

**Comment by Eikju** [comment]
Ah, I saw it right now xDDD

**Comment by Eikju** [comment]
Hah. Now it works. Damn... I do not have a compiler so searching for errors before activate the addon seems very odd xD

**Comment by NielsH** [comment]
Sublime editor will show you syntax errors at least, though if you have typos it'll have to show up in actual testing :p 

Anyways, looks good. There's a few things that aren't needed (which are only used internally in the addon and are never printed, so no need to translate) but I can filter them out :) 

Be careful you don't give your msg strings too much text - It'll still have to fit in the window itself. Also I'm not sure if %s will be picked up properly when you try to format it like that. We'll see when you test it though! :)

**Comment by Eikju** [comment]
Ok. As I mentioned before, I have never programmed anything like that. Neither C+, C# or Lua. Hopefully it will work :D

**Comment by olbat** [comment]
Hi,
I've made a small -dirty- script to translate RaidCore in French for our guild, it also has a German support so it may be useful here as well :) See #15 to get more details about the script.

**Comment by Eikju** [comment]
Hey olbat!

I all ready translated the whole GA in German. With the message strings it sould be very easy to translate the addon into other languages like French. To find out the name of a NPC for English, French or German is very easy. The main problems are skillnames or messages from the boss himself. If you solve these "problems", the addon will be working perfectly :)

**Comment by olbat** [comment]
I agree, finding translations is easy but it's a pain to do it by hand, that's pretty much why I've written this script :)

For the moment, the script handles the skill names ~ correctly but not the messages sent by the boss.
I'm trying to figure out how to automate this, but since this messages are not referenced in any online database, I think that there is no other choices than to do it by hand ... :(

**Comment by olbat** [comment]
If you are interested by taking a look at the changes made by the script on the module's sources (spell names, ...), here is the diff: https://github.com/olbat/RaidCore/commit/635547cc6e73bfd6abf2ba7eb1510870e6e64f8b .

**Comment by Eikju** [comment]
Just for your information @NielsH : My translation worked perfectly. So you can put it into the curse client :)

**Comment by NielsH** [comment]
@Wobel have you had a look at this by any chance to see if it worked well for you as well in GA? I'm a bit reluctant merging it into the master since it's a significant change and I'd rather not break stuff for Curse users!

**Comment by Gil-B** [comment]
I ran with it last night and it seemed ok.
Just need to add translations for the event objectives that were added
after the version that was translated :)

**Comment by NielsH** [comment]
I've added a new translation branch with all the GA translations + current master. See https://github.com/NielsH/RaidCore/tree/translation2

@Eikju - The new master has some restrictions for events/subzones, this is so we can better detect which boss we are currently facing. Could you translate those to english and add them? For example, see https://github.com/NielsH/RaidCore/blob/translation2/Modules/Ohmna.lua#L11 and https://github.com/NielsH/RaidCore/blob/translation2/Modules/Phagemaw.lua#L11

Cheers :)

**Comment by Eikju** [comment]
Sure :D. I will do it for you m8 :)

**Comment by Eikju** [comment]
Done ;-)

**Comment by NielsH** [comment]
Where's the magic stuff? :)

**Comment by Eikju** [comment]
hmm... I pushed the edit button. So is there no notification?

e. g.: https://github.com/NielsH/RaidCore/commit/c0d4e415acefb5cb7af309f8fb6fd9adb4d6545e

**Comment by NielsH** [comment]
I can see that specific commit when I click it directly, but that is not merged into the branch itself. Can you do a pull request for them into the translation2 branch?

Or just link me the gist files and I'll add it :)

**Comment by Eikju** [comment]
Can you see it now? :D

**Comment by NielsH** [comment]
Well I just got 6 mails with them.. :P

**Comment by Eikju** [comment]
Well... I made pull requests which can me automatically merged by you. ^^

https://github.com/NielsH/RaidCore/pull/21
https://github.com/NielsH/RaidCore/pull/20
https://github.com/NielsH/RaidCore/pull/19
https://github.com/NielsH/RaidCore/pull/18
https://github.com/NielsH/RaidCore/pull/22
https://github.com/NielsH/RaidCore/pull/23

**Comment by NielsH** [comment]
I'm gonna need you to do a pull request for the translation2 branch though - not the master branch like you have done now ;)

Almost!

**Comment by Eikju** [comment]
Second try ;-)

https://github.com/NielsH/RaidCore/pull/24
https://github.com/NielsH/RaidCore/pull/25
https://github.com/NielsH/RaidCore/pull/26
https://github.com/NielsH/RaidCore/pull/27
https://github.com/NielsH/RaidCore/pull/28
https://github.com/NielsH/RaidCore/pull/29

**Comment by NielsH** [comment]
Cool, merged them in. Make sure to try next reset if it activates properly. The Subzone/Event restrictions are new, and haven't been tested yet properly with multiple conditions.

I'll run it as well next reset. If all goes well, i'll merge it with master so it makes the next curse release :)

**Comment by Eikju** [comment]
OK. Thanks for that :D

**Comment by Gil-B** [comment]
I'll run it too :)

**Comment by NielsH** [comment]
Note/Reminder to self: We should rename piliers to pillars :p 
https://github.com/NielsH/RaidCore/blob/translation2/Modules/PhageCouncil.lua
https://github.com/NielsH/RaidCore/blob/translation2/Modules/Ohmna.lua

**Comment by Immemor** [comment]
Hi , I didn't read all  message on this "issue", but I think it's the same as https://github.com/NielsH/RaidCore/issues/15. Isn't it ?
